### PR TITLE
WB-1676.2: Add a11y features to announce combobox changes to screen readers

### PR DESCRIPTION
**Pull request by jandrade**
## Summary:

This PR focuses on bringing a better accessibility experience by adding a live region to announce the different combobox statuses:
- When the visual focus changes in the listbox.
- When the visual focus changes in the selected options (multi-select mode).
- When it is closed.
- Describing a state for the focused option. e.g. `selected`, `disabled`.

### Combobox Implementation Plan:

1. #2216
2. #2221
3. #2223
4. **Improve accessibility support on Combobox (labels, aria-live).[CURRENT]**
5. Add autocomplete support to `Combobox` component.
6. Add async/dynamic support to `Combobox` component.

Issue: https://khanacademy.atlassian.net/browse/WB-1676

## Test plan:

Turn on Voice Over and use Safari:

### Single Selection:

1. Navigate to: /iframe.html?args=&id=packages-dropdown-combobox--single-select-combobox&viewMode=story
2. Using the keyboard, open the listbox by focusing on the input then pressing `arrow down`.
3. Verify that the Screen Reader announces the selected item (`Pear selected, 3 of 9. 9 results available`).
4. Navigate on the listbox by using the arrow keys.
5. Verify that it announces the currently focused item (e.g. `Strawberry disabled, 2 of 9. 9 results available`).
6. Using the keyboard, Select one of the enabled options.
7. Verify that it announces the selection (e.g. `Apple selected`) and the listbox is closed.

https://github.com/user-attachments/assets/dd7f6173-e00b-4a96-bed3-c5b0df6c2fbd

### Multiple Selection:

1. Navigate to /iframe.html?args=&id=packages-dropdown-combobox--multiple-selection&viewMode=story
2. Using the keyboard, open the listbox by focusing on the input then pressing `arrow down`.
3. Verify that the Screen Reader announces the selected item (`Pear selected, 3 of 9. 9 results available`).
4. Navigate on the listbox by using the arrow keys.
5. Verify that it announces the currently focused item (e.g. `Strawberry disabled, 2 of 9. 9 results available`).
6. Using the keyboard, Select one of the enabled options (pressing `Enter` on the focused option).
7. Verify that it announces the selection (e.g. `Apple selected, 5 of 9. 9 results available.`) and the listbox remains open).
9. Press arrow left or arrow right to move visual focus to the selected pills.
10. Verify that the focus moves to one of the pills and SR announces that (e.g. `Grape focused, 3 of 3. 3 selected options`).

https://github.com/user-attachments/assets/80a6eabc-ef82-4828-8c18-ed8e05bf8fee

### CODE REVIEW

File: packages/wonder-blocks-dropdown/src/components/combobox-live-region.tsx [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+            const selectedState =
+                selectedLength > lastSelectedLength
+                    ? "selected"
+                    : "not selected";
+            // Multi-select combobox.
+            if (Array.isArray(selected) && selected.length > 0) {
+                newMessage = `${multiSelectLabels.join(", ")} ${selectedState}`;
+            } else {
+                // Announces the selected item for single select combobox.
+                newMessage = multiSelectLabels[0] + " selected";
```
Comment (beaesguerra) [inline packages/wonder-blocks-dropdown/src/components/combobox-live-region.tsx:105]: Should we get the selected label from the options and selected props? From the prop description for `multiSelectLabels`, it seems like it should only be used for the multiselect!

File: packages/wonder-blocks-dropdown/src/components/combobox-live-region.tsx [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            }
+            setMessage(newMessage);
+        }
+
+        lastSelectedValue.current = selected;
+
+        if (selectionType === "multiple" && !opened) {
+            setMessage(labels.closedState);
```
Comment (beaesguerra) [inline packages/wonder-blocks-dropdown/src/components/combobox-live-region.tsx:113]: Nice add! It's too bad that the browsers don't have built in announcements for the expanded/closed states for comboboxes based on aria-expanded 😢

File: packages/wonder-blocks-dropdown/src/components/__tests__/combobox-live-region.test.tsx [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                    multiSelectLabels={[]}
+                    opened={true}
+                />,
+            );
+
+            // Assert
+            expect(screen.getByRole("log")).toHaveTextContent("");
+        });
+
+        // it.todo("announces when we move the focus to the next pill");
```
Comment (beaesguerra) [inline packages/wonder-blocks-dropdown/src/components/__tests__/combobox-live-region.test.tsx:82]: nit: Did we want to keep this todo around?

**Review by beaesguerra — COMMENTED** [review]
Nice a11y improvements 🎉 ! I left some questions 😄 

Do we also want to document these a11y features in Storybook?

**Comment by beaesguerra** [comment]
Also, unrelated to the a11y features, I noticed that in some of the stories (like the [Combobox with single selection](https://5e1bf4b385e3fb0020b7073c-mfdbnbdldi.chromatic.com/?path=/story/packages-dropdown-combobox--single-select-combobox)), the field is showing the option value instead of the label (`pear` vs `Pear`)! Not sure if this will be addressed in one of the future PRs!

File: packages/wonder-blocks-dropdown/src/components/combobox-live-region.tsx [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+    // Announce when an item is selected.
+    React.useEffect(() => {
+        if (selected !== lastSelectedValue?.current) {
+            let newMessage = "";
+            const lastSelectedLength = lastSelectedValue?.current?.length ?? 0;
+            const selectedLength = selected?.length ?? 0;
+            const selectedState =
+                selectedLength > lastSelectedLength
+                    ? "selected"
```
[the remark this answers is not in the document]
Comment (jandrade) [reply packages/wonder-blocks-dropdown/src/components/combobox-live-region.tsx]: Good point! I added that to `labels`.

File: packages/wonder-blocks-dropdown/src/components/combobox-live-region.tsx (hunk 2 above)
Comment (jandrade) [reply packages/wonder-blocks-dropdown/src/components/combobox-live-region.tsx:120]: So true

**Comment by jandrade** [comment]
> Also, unrelated to the a11y features, I noticed that in some of the stories (like the [Combobox with single selection](https://5e1bf4b385e3fb0020b7073c-mfdbnbdldi.chromatic.com/?path=/story/packages-dropdown-combobox--single-select-combobox)), the field is showing the option value instead of the label (`pear` vs `Pear`)! Not sure if this will be addressed in one of the future PRs!

Ooh @beaesguerra good catch! I'll fix that in a separate PR (also have to fix the border colors to match the rest of form fields), so I'll group both things together.

File: __docs__/wonder-blocks-dropdown/combobox-accessibility.mdx [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+<Meta title="Packages / Dropdown / Combobox / Accessibility" />
+
+# Combobox Accessibility
+
+The Combobox component is built with accessibility in mind. It includes the
+following features (as advised by the <a
+href="https://www.w3.org/WAI/ARIA/apg/patterns/combobox/">WAI-ARIA Authoring
+Practices</a>):
+
+## Keyboard interaction
```
Comment (beaesguerra) [inline __docs__/wonder-blocks-dropdown/combobox-accessibility.mdx:12]: Nice! These docs are helpful in knowing what the implemented interactions are 👍

File: __docs__/wonder-blocks-dropdown/combobox-accessibility.mdx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+-   `ArrowDown`: Moves visual focus to the next option. If the visual focus is
+    on the last option, the visual focus moves to the first option (cyclic
+    navigation).
+-   `ArrowUp`: Moves visual focus to the previous option. If the visual focus in
+    on the first option, the visual focus moves to the last option (cyclic
+    navigation).
+-   `Home`: Moves visual focus to the first option.
+-   `End`: Moves visual focus to the last option.
+-   `Enter`: Sets the Textbox/combobox value to the content of the focused
+    option in the listbox, then closes the listbox.
```
Comment (beaesguerra) [inline __docs__/wonder-blocks-dropdown/combobox-accessibility.mdx:34]: nit: The listbox doesn't close when `Enter` is pressed when it is a multiselect! Something like:
```suggestion
    option in the listbox, then closes the listbox if only one selection can be made.
```

File: __docs__/wonder-blocks-dropdown/combobox-accessibility.mdx [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+The Combobox component is designed to be screen reader friendly.
+
+We provide a custom live region that will announce changes to the combobox due
+to some limitations in browsers (e.g. Safari). This live region will announce
+the following:
+
+-   When the listbox is opened, the focused option in the listbox, including the
+    index of the option and the total number of options (e.g. `Pear selected, 3
+of 9. 9 results available`).
+-   When the user tabs through the listbox, the focused option in the listbox,
```
Comment (beaesguerra) [inline __docs__/wonder-blocks-dropdown/combobox-accessibility.mdx:53]: nit: The options in the listbox aren't tabbable so this sentence could cause confusion!

```suggestion
-   When the user navigates through the listbox, the focused option in the listbox,
```

**Review by beaesguerra — APPROVED** [review]
Looks great 🎉  The accessibility docs are thorough and helpful in communicating the behaviour/specs, thanks for adding that in!

File: __docs__/wonder-blocks-dropdown/combobox-accessibility.mdx (hunk 6 above)
Comment (jandrade) [reply __docs__/wonder-blocks-dropdown/combobox-accessibility.mdx:34]: good point! I add your recommendation.

**Comment by jandrade** [comment]
The parent pull-request ([#2223](https://github.com/Khan/wonder-blocks/pull/2223)) has been merged into `main`, but this branch (`combobox-a11y`) now has conflicts with the new base branch. These conflicts must be resolved before checks can complete on this pull-request.

**Pull request merged**